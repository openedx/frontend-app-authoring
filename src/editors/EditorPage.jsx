import React from 'react';
import PropTypes from 'prop-types';
import { useParams } from 'react-router';
import Placeholder from '@edx/frontend-lib-content-components';

const EditorPage = ({
    courseId,
}) =>{
    const params = useParams();
    const blockType = params.blockType,
        blockId = params.blockId;
    return (
        <div className = "editor-page">
            <Placeholder/>
        </div>
    );
}
EditorPage.propTypes = {
    courseId: PropTypes.string.isRequired,
  };

export default EditorPage;
