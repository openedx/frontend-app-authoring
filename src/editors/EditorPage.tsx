import React from 'react';
import { Provider } from 'react-redux';

import store from './data/store';
import Editor from './Editor';
import ErrorBoundary from './sharedComponents/ErrorBoundary';
import { EditorComponent } from './EditorComponent';
import { EditorContextProvider } from './EditorContext';

interface Props extends EditorComponent {
  blockId?: string;
  blockType: string;
  courseId: string;
  lmsEndpointUrl?: string;
  studioEndpointUrl?: string;
  fullScreen?: boolean;
  children?: never;
}

/**
 * Wraps the editors with the redux state provider.
 * TODO: refactor some of this to be React Context and React Query
 */
const EditorPage: React.FC<Props> = ({
  courseId,
  blockType,
  blockId = null,
  lmsEndpointUrl = null,
  studioEndpointUrl = null,
  onClose = null,
  returnFunction = null,
  fullScreen = true,
}) => (
  <Provider store={store}>
    <ErrorBoundary
      {...{
        learningContextId: courseId,
        studioEndpointUrl,
      }}
    >
      <EditorContextProvider fullScreen={fullScreen} learningContextId={courseId}>
        <Editor
          {...{
            onClose,
            learningContextId: courseId,
            blockType,
            blockId,
            lmsEndpointUrl,
            studioEndpointUrl,
            returnFunction,
          }}
        />
      </EditorContextProvider>
    </ErrorBoundary>
  </Provider>
);

export default EditorPage;
