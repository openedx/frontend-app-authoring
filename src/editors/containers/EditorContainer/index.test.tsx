import { snakeCaseObject } from '@edx/frontend-platform';
import {
  render,
  screen,
  initializeMocks,
  fireEvent,
  waitFor,
} from '../../../testUtils';
import editorCmsApi from '../../data/services/cms/api';

import EditorPage from '../../EditorPage';

// Mock this plugins component:
jest.mock('frontend-components-tinymce-advanced-plugins', () => ({ a11ycheckerCss: '' }));
// Always mock out the "fetch course images" endpoint:
jest.spyOn(editorCmsApi, 'fetchImages').mockImplementation(async () => ( // eslint-disable-next-line
  { data: { assets: [], start: 0, end: 0, page: 0, pageSize: 50, totalCount: 0 } }
));
// Mock out the 'get ancestors' API:
jest.spyOn(editorCmsApi, 'fetchByUnitId').mockImplementation(async () => ({
  status: 200,
  data: {
    ancestors: [{
      id: 'block-v1:Org+TS100+24+type@vertical+block@parent',
      display_name: 'You-Knit? The Test Unit',
      category: 'vertical',
      has_children: true,
    }],
  },
}));

const defaultPropsHtml = {
  blockId: 'block-v1:Org+TS100+24+type@html+block@123456html',
  blockType: 'html',
  courseId: 'course-v1:Org+TS100+24',
  lmsEndpointUrl: 'http://lms.test.none/',
  studioEndpointUrl: 'http://cms.test.none/',
  onClose: jest.fn(),
  fullScreen: false,
};
const fieldsHtml = {
  displayName: 'Introduction to Testing',
  data: '<p>This is a text component which uses <strong>HTML</strong>.</p>',
  metadata: { displayName: 'Introduction to Testing' },
};

describe('EditorContainer', () => {
  beforeEach(() => {
    initializeMocks();
  });

  test('it displays a confirmation dialog when closing the editor modal', async () => {
    jest.spyOn(editorCmsApi, 'fetchBlockById').mockImplementationOnce(async () => (
      { status: 200, data: snakeCaseObject(fieldsHtml) }
    ));

    render(<EditorPage {...defaultPropsHtml} />);

    // Then the editor should open
    expect(await screen.findByRole('heading', { name: /Introduction to Testing/ })).toBeInTheDocument();

    // Assert the "are you sure?" message isn't visible yet
    const confirmMessage = /Are you sure you want to exit the editor/;
    expect(screen.queryByText(confirmMessage)).not.toBeInTheDocument();

    // Find and click the close button
    const closeButton = await screen.findByRole('button', { name: 'Exit the editor' });
    fireEvent.click(closeButton);
    // Now we should see the confirmation message:
    expect(await screen.findByText(confirmMessage)).toBeInTheDocument();

    expect(defaultPropsHtml.onClose).not.toHaveBeenCalled();
    // And can confirm the cancelation:
    const confirmButton = await screen.findByRole('button', { name: 'OK' });
    fireEvent.click(confirmButton);
    expect(defaultPropsHtml.onClose).toHaveBeenCalled();
  });

  test('it disables the save button until the fields have been loaded', async () => {
    // Mock that loading the block data has begun but not completed yet:
    let resolver: (result: { data: any }) => void;
    jest.spyOn(editorCmsApi, 'fetchBlockById').mockImplementationOnce(() => new Promise((r) => { resolver = r; }));

    render(<EditorPage {...defaultPropsHtml} />);

    // Then the editor should open. The "Save" button should be disabled
    const saveButton = await screen.findByRole('button', { name: /Save changes and return/ });
    expect(saveButton).toBeDisabled();

    // Now complete the loading of the data:
    await waitFor(() => expect(resolver).not.toBeUndefined());
    resolver!({ data: snakeCaseObject(fieldsHtml) });

    // Now the save button should be active:
    await waitFor(() => expect(saveButton).not.toBeDisabled());
  });
});
