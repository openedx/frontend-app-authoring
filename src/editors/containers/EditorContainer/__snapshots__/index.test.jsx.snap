// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorContainer component render snapshot: initialized. enable save and pass to header 1`] = `
<div
  className="editor-container d-flex flex-column position-relative zindex-0"
  style={
    {
      "minHeight": "100%",
    }
  }
>
  <BaseModal
    bodyStyle={null}
    close={[MockFunction closeCancelConfirmModal]}
    confirmAction={
      <Button
        onClick={[Function]}
        variant="primary"
      >
        <FormattedMessage
          defaultMessage="OK"
          description="Label for OK button"
          id="authoring.editorContainer.okButton.label"
        />
      </Button>
    }
    footerAction={null}
    headerComponent={null}
    isFullscreenScroll={true}
    isOpen={false}
    size="md"
    title="Exit the editor?"
  >
    <FormattedMessage
      defaultMessage="Are you sure you want to exit the editor? Any unsaved changes will be lost."
      description="Description text for modal confirming cancellation"
      id="authoring.editorContainer.cancelConfirm.description"
    />
  </BaseModal>
  <ModalDialog.Header
    className="shadow-sm zindex-10"
  >
    <div
      className="d-flex flex-row justify-content-between"
    >
      <h2
        className="h3 col pl-0"
      >
        <injectIntl(ShimmedIntlComponent)
          isInitialized={true}
        />
      </h2>
      <IconButton
        alt="Exit the editor"
        iconAs="Icon"
        onClick={[MockFunction openCancelConfirmModal]}
        src={[MockFunction icons.Close]}
      />
    </div>
  </ModalDialog.Header>
  <ModalDialog.Body
    className="pb-0 mb-6"
  >
    <h1>
      My test content
    </h1>
  </ModalDialog.Body>
  <injectIntl(ShimmedIntlComponent)
    disableSave={false}
    onCancel={[MockFunction openCancelConfirmModal]}
    onSave={
      {
        "handleSaveClicked": {
          "dispatch": [MockFunction react-redux.dispatch],
          "getContent": [MockFunction props.getContent],
          "returnFunction": [MockFunction props.returnFunction],
          "validateEntry": [MockFunction props.validateEntry],
        },
      }
    }
  />
</div>
`;

exports[`EditorContainer component render snapshot: not initialized. disable save and pass to header 1`] = `
<div
  className="editor-container d-flex flex-column position-relative zindex-0"
  style={
    {
      "minHeight": "100%",
    }
  }
>
  <BaseModal
    bodyStyle={null}
    close={[MockFunction closeCancelConfirmModal]}
    confirmAction={
      <Button
        onClick={[Function]}
        variant="primary"
      >
        <FormattedMessage
          defaultMessage="OK"
          description="Label for OK button"
          id="authoring.editorContainer.okButton.label"
        />
      </Button>
    }
    footerAction={null}
    headerComponent={null}
    isFullscreenScroll={true}
    isOpen={false}
    size="md"
    title="Exit the editor?"
  >
    <FormattedMessage
      defaultMessage="Are you sure you want to exit the editor? Any unsaved changes will be lost."
      description="Description text for modal confirming cancellation"
      id="authoring.editorContainer.cancelConfirm.description"
    />
  </BaseModal>
  <ModalDialog.Header
    className="shadow-sm zindex-10"
  >
    <div
      className="d-flex flex-row justify-content-between"
    >
      <h2
        className="h3 col pl-0"
      >
        <injectIntl(ShimmedIntlComponent)
          isInitialized={false}
        />
      </h2>
      <IconButton
        alt="Exit the editor"
        iconAs="Icon"
        onClick={[MockFunction openCancelConfirmModal]}
        src={[MockFunction icons.Close]}
      />
    </div>
  </ModalDialog.Header>
  <ModalDialog.Body
    className="pb-0 mb-6"
  />
  <injectIntl(ShimmedIntlComponent)
    disableSave={true}
    onCancel={[MockFunction openCancelConfirmModal]}
    onSave={
      {
        "handleSaveClicked": {
          "dispatch": [MockFunction react-redux.dispatch],
          "getContent": [MockFunction props.getContent],
          "returnFunction": [MockFunction props.returnFunction],
          "validateEntry": [MockFunction props.validateEntry],
        },
      }
    }
  />
</div>
`;
